### ISSUE TITLE
Optimize geohash/geohashes-intersecting faster

### ISSUE THREAD

**Issue by chen-factual** [Factual/geo#25]
It takes ~5 minutes on a beefy machine to split the US polygon into 778 L3 geohashes. This seems a bit ridiculous.

Profiling the job, I see we're burning 92% of compute time in `org.locationtech.spatial4j.shape.jts.JtsShapeFactory.makeShape()`, so looking at how we can reduce calls there seems like an obvious place to optimize.

**Comment by worace** [comment]
Dang if that is true we should definitely look at it; can you share the polygon you are using? I am not seeing it that bad using this as a test:

```
{
    "type": "MultiPolygon",
    "coordinates":
    [
        [
            [
                [ -123.123779, 48.227039 ],  
                [ -123.318787, 49.000042 ],
                [ -121.742592, 49.000267 ],
                [ -95.157394, 49.000493 ],
                [ -95.157394, 49.390418 ],
                [ -94.795532, 49.357334 ],
                [ -94.482422, 48.857487 ],
                [ -88.36853, 48.314255 ],
                [ -84.126389, 46.531937 ],
                [ -81.331787, 45.344424 ],
                [ -83.034668, 41.910453 ],
                [ -79.013672, 42.867912 ],
                [ -79.299316, 43.590338 ],
                [ -77.305298, 43.761176 ],
                [ -74.849854, 45.058001 ],
                [ -71.586914, 45.1123 ],
                [ -69.213867, 47.480088 ], 
                [ -67.758179, 47.271775 ],
                [ -67.719727, 45.813486 ],
                [ -66.780396, 44.785734],
                [ -80.628662, 24.417142 ],
                [ -97.058716, 25.730633 ],
                [ -99.283447, 26.382028 ],
                [ -101.480713, 29.678508 ],
                [ -102.612305, 29.716681 ],
                [ -103.117676, 28.88316 ],
                [ -104.699707, 29.649869 ],
                [ -106.44104, 31.737511 ],
                [ -108.187866, 31.760867 ],
                [ -108.193359, 31.325487 ],
                [ -111.08551, 31.325487 ],
                [ -114.930725, 32.521342 ],
                [ -114.724731, 32.711044 ],
                [ -124.892578, 31.952453 ],
                [ -129.067383, 49.047486 ],
                [ -123.123779, 48.227039 ]
            ] 
        ], 
        [ 
            [ 
                [ -163.916016, 71.992578 ], 
                [ -140.888672, 70.641769 ],
                [ -140.976562, 60.326948 ],
                [ -135.175781, 60.326948 ],
                [ -129.550781, 55.553495 ],
                [ -131.286621, 54.239551 ],
                [ -179.736328, 51.069017 ],
                [ -172.089844, 63.626745 ],
                [ -163.916016, 71.992578 ]
            ]
        ],
        [ 
            [ 
                [ -161.323242, 22.512557 ], 
                [ -152.446289, 22.065278 ],
                [ -156.09375, 17.811456 ],
                [ -161.323242, 22.512557 ]
            ]
        ]
    ]
}
```

```
(time (count (gh/geohashes-intersecting (gio/read-geojson (slurp "/tmp/usa.geojson")) 15)))
"Elapsed time: 40.235755 msecs"
```

**Comment by worace** [comment]
OTOH this impl is definitely slower than the one in `geoq`

```
(time (count (gh/geohashes-intersecting (gio/read-geojson (slurp "/tmp/usa.geojson")) 25)))
"Elapsed time: 22501.496688 msecs"
899817
```
vs

```
time cat /tmp/usa.geojson|jq -cr . | geoq gh covering 5 | wc -l
899817
cat /tmp/usa.geojson  0.00s user 0.00s system 45% cpu 0.001 total
noglob jq -cr .  0.00s user 0.00s system 89% cpu 0.003 total
geoq gh covering 5  2.25s user 0.42s system 99% cpu 2.667 total
wc -l  0.13s user 0.42s system 20% cpu 2.666 total
```

**Comment by worace** [comment]
@chen-factual pushed 2.0.0-rc-3

**Comment by chen-factual** [comment]
PR submitted, didn't get to clean up as much as I wanted to. To clarify my above comment my benchmark was splitting into L3s not L7s.

One thing I noticed was subdivide was killing a lot of time in lazy-seqs. This only really show up when I was looking for smaller geohashes. I'll routinely be doing some L2 -> L7 splits so maybe there is some attention to pay there as well, but I'll do some basic benchmarks before filing that ticket.

**Comment by saifalharthi** [comment]
Hi ,
I am interested in looking at this problem (and learning Clojure as well). However, when I tried re-running the code to benchmark the current code I got this error:
```
user=> (require '[geo [geohash :as geohash] [poly :as poly] [jts :as jts] [spatial :as spatial] [io :as gio]])
nil
user=> (time (count (geohash/geohashes-intersecting (gio/read-geojson (slurp "tmp/USA.geojson")) 25)))

IllegalArgumentException No implementation of method: :to-shape of protocol: #'geo.spatial/Shapelike found for class: clojure.lang.LazySeq  clojure.core/-cache-protocol-fn (core_deftype.clj:583)
```
I would note that I tried it both by installing the library in a new project and from source. Can you tell me where I went wrong?

**Comment by worace** [comment]
Hi @saifalharthi sorry about that; I realize now that in version 2.0 we changed the way that function works a little bit and forgot to update the example in the readme. I just updated the example, but the short version is that `read-geojson` now returns a collection rather than a single geometry or feature (this is done in order to make it possible to read geojson FeatureCollections). There is also a function `read-geojson-geometry` which does what `read-geojson` used to do, i.e. return a JTS geometry from a single GeoJSON element.

For more info you might like to check out the information in the API documentation here: http://factual.github.io/geo/2.1.0/geo.io.html. Those have a little more explanation than the examples in the readme.

hope that helps!
